### PR TITLE
Refactor Objects in ViewTrips Directory

### LINKED ISSUE #84: Switch Objects in ViewTrips Direc. from Struct to Dict
### Issue summary

Consider changing all objects that have keys corresponding to the constants in `database.js` that currently use the struct syntax to use dict syntax (see [here](https://github.com/google/closure-compiler/wiki/%40struct-and-%40dict-Annotations) for details).

Note: be careful to change every one because some of the naming for fields is different with the struct syntax than the constants in `database.js` (e.g. startData rather than start_date).

### Actual behavior

Instead of declaring and accessing like so:
```
const tripData = {  
    name: ...,
    description: ...,
};
...
const name = tripData.name;
```

### Expected behavior

Do this (where DB is the imported `database.js`):
```
let tripData = {};
tripData[DB.TRIPS_NAME] = ...;
...
const name = tripData[DB.TRIPS_NAME];
```

### PR DESCRIPTION
**Pull request by zghera**
### What is a quick description of the change?

Updates all JS objects to use the dict rather than struct format for updating/accessing in order to use the constants in `database.js` in all appropriate cases.

### Is this fixing an issue?

fixes #84

### Are there more details that are relevant?

Updated fields in `database.js` for more clarity:
- `TRIPS_NAME` --> `TRIPS_TITLE`
- `TRIPS_CREATION_TIME` --> `TRIPS_UPDATE_TIMESTAMP`

Also had to create a new [firestore index](https://firebase.corp.google.com/u/0/project/step53-2020/database/firestore/indexes?create_composite=Cklwcm9qZWN0cy9zdGVwNTMtMjAyMC9kYXRhYmFzZXMvKGRlZmF1bHQpL2NvbGxlY3Rpb25Hcm91cHMvdHJpcHMvaW5kZXhlcy9fEAEaEQoNY29sbGFib3JhdG9ycxgBGhkKFXRyaXBfdXBkYXRlX3RpbWVzdGFtcBACGgwKCF9fbmFtZV9fEAI) as a result.

### Check lists (check `x` in `[ ]` of list items)

- [x] Test written/updating
- [x] Tests passing
- [x] Coding style (indentation, etc)

### CODE REVIEW

File: frontend/src/components/ViewTrips/trip.js [hunk 1]
```diff
@@ -56,26 +57,32 @@ export function getCollaboratorEmails(collaboratorUidArr) {
...
 const Trip = (props) => {
-  const name = props.tripData.name;
-  const description = props.tripData.description;
-  const destination = props.tripData.destination;
+  // Unpack trip document data.
+  const title = props.tripData[DB.TRIPS_TITLE];
+  const description = props.tripData[DB.TRIPS_DESCRIPTION];
+  const destination = props.tripData[DB.TRIPS_DESTINATION];
+  const startDateTimestamp = props.tripData[DB.TRIPS_START_DATE];
+  const endDateTimestamp = props.tripData[DB.TRIPS_END_DATE];
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/trip.js:65]: Are these timestamp objects? just curious.

File: frontend/src/components/ViewTrips/trip.js [hunk 2]
```diff
@@ -6,24 +6,25 @@ import { timestampToISOString } from '../Utils/time.js';
...
+ * @param {!firebase.firestore.Timestamp} startDateTimestamp Firestore timestamp
+ *     Object corresponding to the trip start date.
+ * @param {!firebase.firestore.Timestamp} endDateTimestamp Firestore timestamp
+ *     Object corresponding to the trip end date.
+ * @return {string} Date range of the trip in the form 'MM/DD/YYY - MM/DD/YYY'.
  */
-export function getDateRange(tripData) {
-  const startDate = tripData.start_date.toDate();
-  const endDate = tripData.end_date.toDate();
+export function getDateRange(startDateTimestamp, endDateTimestamp) {
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/trip.js:25]: this function is another friend to time.js, if you'd like to share :)

File: frontend/src/components/ViewTrips/trip.test.js [hunk 3]
```diff
@@ -14,12 +14,12 @@ test('getDateRange test', () => {
...
-  const testStartDate = firebase.firestore.Timestamp.fromDate(
+  const testStartDateTimestamp = firebase.firestore.Timestamp.fromDate(
       new Date(startYear, startMonth - 1, startDay));
-  const testEndDate = firebase.firestore.Timestamp.fromDate(
+  const testEndDateTimestamp = firebase.firestore.Timestamp.fromDate(
       new Date(endYear, endMonth - 1, endDay));
-  const testTripObj = {start_date: testStartDate, end_date: testEndDate};
-  const testDateRange = getDateRange(testTripObj);
+  const testDateRange = getDateRange(testStartDateTimestamp,
+                                     testEndDateTimestamp);
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/trip.test.js:22]:
```suggestion
  const testDateRange = getDateRange(testStartDateTimestamp, testEndDateTimestamp);
```
feel free to ignore this suggestion

File: frontend/src/components/ViewTrips/trips-container.js [hunk 4]
```diff
@@ -18,13 +18,13 @@ const db = app.firestore();
...
     console.log(`Error in Trips Container: ${error}`);
-    resolve(
+    resolve((
       <div>
         <p>Oops, it looks like we were unable to load your trips.
                       Please wait a few minutes and try again.
         </p>
       </div>
-    );
+  ));
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/trips-container.js:27]: Why the double parentheses (green highlighted change)?

File: frontend/src/components/ViewTrips/trips-container.js [hunk 5]
```diff
@@ -72,8 +73,8 @@ class TripsContainer extends React.Component {
           );
 
           this.setState({ trips: tripsContainer });
-        }, (error) => {
-          let errorElement = getErrorElement(error);
+        }, async (error) => {
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/trips-container.js:76]: Why do you need the async tag here? the whole function is asynchronous by nature.

File: frontend/src/components/ViewTrips/trip.js [hunk 6]
```diff
@@ -6,24 +6,25 @@ import { timestampToISOString } from '../Utils/time.js';
...
  * Returns the string date range of the trip associated with the Trip document
- * data `tripObj`.
+ * start and end date timestamps.
  *
- * Notes:
- *  - tripObj will always contain valid start_date and end_date fields.
- *  - When the Firestore Timestamps contained in `tripObj` converted to js
+ * Note: When the Firestore Timestamps contained in `tripObj` converted to js
  *    dates, the months are 0 indexed rather than 1 indexed so they must be
  *    incremented by 1 in order for the month to be correct.
```
[the remark this answers is not in the document]
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:17]: It was a pretty beefy comment compared to the function size so I thought it made sense to place it in the JSDoc. I don't think there is that big of a difference so I will keep it as is unless you really think that it belongs inside. Just let me know.

File: frontend/src/components/ViewTrips/trip.js [hunk 7]
```diff
@@ -6,24 +6,25 @@ import { timestampToISOString } from '../Utils/time.js';
...
  *
- * @param {!firebase.firestore.DocumentData} tripData Object containing the
- *     fields and values for a Trip document.
- * @return {string} Date range of the trip.
+ * @param {!firebase.firestore.Timestamp} startDateTimestamp Firestore timestamp
+ *     Object corresponding to the trip start date.
+ * @param {!firebase.firestore.Timestamp} endDateTimestamp Firestore timestamp
+ *     Object corresponding to the trip end date.
+ * @return {string} Date range of the trip in the form 'MM/DD/YYY - MM/DD/YYY'.
```
[the remark this answers is not in the document]
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:23]: Ahh yes, the good ol' days back in the 20s. The real 20s not the 1920s. Fixed.

File: frontend/src/components/ViewTrips/trip.js [hunk 8]
```diff
@@ -6,24 +6,25 @@ import { timestampToISOString } from '../Utils/time.js';
...
  */
-export function getDateRange(tripData) {
-  const startDate = tripData.start_date.toDate();
-  const endDate = tripData.end_date.toDate();
+export function getDateRange(startDateTimestamp, endDateTimestamp) {
+  const startDate = startDateTimestamp.toDate();
+  const endDate = endDateTimestamp.toDate();
   return `${startDate.getMonth() + 1}/${startDate.getDate()}/`  +
       `${startDate.getFullYear()} - ${endDate.getMonth() + 1}/` +
       `${endDate.getDate()}/${endDate.getFullYear()}`;
```
[the remark this answers is not in the document]
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:30]: Oh nice I did not know about that! I switched to using that.

File: frontend/src/components/ViewTrips/trip.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:17]: No longer applicable.

File: frontend/src/components/ViewTrips/trip.js (hunk 1 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:63]: Yes

File: frontend/src/components/ViewTrips/trip.js (hunk 2 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:25]: Yeah I will plop it in there even though it is pretty simple at this point.

File: frontend/src/components/ViewTrips/trip.test.js (hunk 3 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.test.js]: No longer applicable.

File: frontend/src/components/ViewTrips/trips-container.js (hunk 4 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trips-container.js]: Whoops this was an accident. Must have done this when I ran into a bug of not calling this function with async. Changed it back.

File: frontend/src/components/ViewTrips/trips-container.js (hunk 5 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trips-container.js:76]: I thought so too and I had a bug because of it. I would never have seen this if I didn't change the query which required me to create a database index (which caused the error callback to execute).

So because of the structure of `onSnapshot` you have to pass a callback for the error function. Even though componentDidMount is async, it seems that for some reason that if I do not explicitly return a promise from `getErrorElement`, the Promise world gets mad. So in order to use the `await` keyword for `getErrorElement`, I have to make the function async.

File: frontend/src/components/Utils/time.js [hunk 9]
```diff
@@ -77,9 +77,30 @@ export function getTimestampFromDateString(dateStr) {
 /**
  * Formats a Firestore timestamp into a date string in ISO format.
  *
+ * No tests for this function because there is no logic.
```
Comment (keiffer01) [inline frontend/src/components/Utils/time.js:80]: Nit: I don't think it's necessary to state whether or not tests for a function exists? I suppose stating it does no harm either but maybe Hiram or Longjun can say more about it.

File: frontend/src/components/Utils/time.js [hunk 10]
```diff
@@ -77,9 +77,30 @@ export function getTimestampFromDateString(dateStr) {
...
 export function timestampToISOString(timestamp) {
   return timestamp.toDate().toISOString().substring(0,10);
 }
+
+
+/**
+ * Returns the string date range of the trip associated with the Trip document
+ * start and end date timestamps.
+ *
+ * No tests for this function because there is no logic.
```
Comment (keiffer01) [inline frontend/src/components/Utils/time.js:94]: As above.

File: frontend/src/components/Utils/time.js [hunk 11]
```diff
@@ -77,9 +77,30 @@ export function getTimestampFromDateString(dateStr) {
...
+ *
+ * @param {!firebase.firestore.Timestamp} startDateTimestamp Firestore timestamp
+ *     Object corresponding to the trip start date.
+ * @param {!firebase.firestore.Timestamp} endDateTimestamp Firestore timestamp
```
Comment (keiffer01) [inline frontend/src/components/Utils/time.js:98]: I must've missed it if you did it before but what is the exclamation marks for?

Comment (zghera) [reply frontend/src/components/Utils/time.js:98]: [Non-null objects for JSDoc](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#the-javascript-type-language)

File: frontend/src/components/Utils/time.js (hunk 9 above)
Comment (zghera) [reply frontend/src/components/Utils/time.js:80]: Yeah I agree this is a little silly. I replaced these comments with some TODOs to remove them when integrating timezones anyway.

File: frontend/src/components/Utils/time.js (hunk 10 above)
Comment (zghera) [reply frontend/src/components/Utils/time.js:94]: As above as well.

File: frontend/src/components/Utils/filter-input.js [hunk 12]
```diff
@@ -58,16 +59,19 @@ export function formatTripData(rawTripObj) {
...
+      getCleanedTextInput(rawTripObj[DB.TRIPS_TITLE], defaultName);
+  tripObj[DB.TRIPS_DESCRIPTION] = rawTripObj[DB.TRIPS_DESCRIPTION];
+  tripObj[DB.TRIPS_DESTINATION] =
+      getCleanedTextInput(rawTripObj[DB.TRIPS_DESTINATION], defaultDestination);
+  tripObj[DB.TRIPS_START_DATE] =
+      getTimestampFromDateString(rawTripObj[DB.TRIPS_START_DATE]);
+  tripObj[DB.TRIPS_END_DATE] =
+      getTimestampFromDateString(rawTripObj[DB.TRIPS_END_DATE]);
+  tripObj[DB.TRIPS_COLLABORATORS] =
+      getCollaboratorUidArray(rawTripObj[DB.TRIPS_COLLABORATORS]);
```
Comment (Hiram-Silvey) [inline frontend/src/components/Utils/filter-input.js:74]: What's the difference between this and just replacing strings like "trip_creation_time" with "DB.TRIPS_UPDATE_TIMESTAMP"?

File: frontend/src/components/ViewTrips/save-trip-modal.js [hunk 13]
```diff
@@ -109,17 +111,16 @@ class SaveTripModal extends React.Component {
...
+    const rawTripData = {};
+    rawTripData[DB.TRIPS_TITLE] = this.titleRef.current.value;
+    rawTripData[DB.TRIPS_DESCRIPTION] = this.descriptionRef.current.value;
+    rawTripData[DB.TRIPS_DESTINATION] = this.destinationRef.current.value;
+    rawTripData[DB.TRIPS_START_DATE] = this.startDateRef.current.value;
+    rawTripData[DB.TRIPS_END_DATE] = this.endDateRef.current.value;
+    rawTripData[DB.TRIPS_COLLABORATORS] =
+        this.state.collaboratorsRefArr.map(ref => ref.current.value);
+
+    const tripData = formatTripData(rawTripData);
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewTrips/save-trip-modal.js:123]: Same question as above.

File: frontend/src/components/ViewTrips/trip.js [hunk 14]
```diff
@@ -56,26 +36,32 @@ export function getCollaboratorEmails(collaboratorUidArr) {
...
+  // SaveTripModal component to use when filling out form input default values.
+  const formattedTripData = {};
+  formattedTripData[DB.TRIPS_TITLE] = title;
+  formattedTripData[DB.TRIPS_DESCRIPTION] = description;
+  formattedTripData[DB.TRIPS_DESTINATION] = destination;
+  formattedTripData[DB.TRIPS_START_DATE] =
+      timestampToISOString(startDateTimestamp);
+  formattedTripData[DB.TRIPS_END_DATE] =
+      timestampToISOString(endDateTimestamp);
+  formattedTripData[DB.TRIPS_COLLABORATORS] = collaboratorEmailsStr.split(', ');
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewTrips/trip.js:58]: Same question as above.

File: frontend/src/components/ViewTrips/trips-container.js [hunk 15]
```diff
@@ -26,7 +26,7 @@ function getErrorElement(error) {
                       Please wait a few minutes and try again.
         </p>
       </div>
-    );
+  );
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewTrips/trips-container.js:29]: Why was this moved? It seemed to be lined up with `resolve` before?

Comment (zghera) [reply frontend/src/components/ViewTrips/trips-container.js:28]: Ohh this was fixed when reverting back to the synchronous version of `getErrorElement`.

File: frontend/src/components/Utils/filter-input.js (hunk 12 above)
Comment (zghera) [reply frontend/src/components/Utils/filter-input.js:73]: After chatting some offline, found that this does not work in ES5 or earlier but there is a way to do it in ES6 with `ComputedPropertyName` syntax (learned about it [here](https://stackoverflow.com/questions/2274242/how-to-use-a-variable-for-a-key-in-a-javascript-object-literal)). Changed to use this.

File: frontend/src/components/Utils/time.test.js [hunk 16]
```diff
@@ -129,6 +129,7 @@ test('ISODate empty input tests', () => {
   expect(utils.getISODate(testDate, null)).toBe(expected);
 });
 
+
```
Comment (Hiram-Silvey) [inline frontend/src/components/Utils/time.test.js]: Remove extra line?

File: frontend/src/components/ViewTrips/trip.js [hunk 17]
```diff
@@ -46,8 +46,28 @@ const Trip = (props) => {
...
+      let collaboratorEmailArr =
+          await authUtils.getUserEmailArrFromUserUidArr(collaboratorUidArr);
+      collaboratorEmailArr = moveCurUserEmailToFront(collaboratorEmailArr);
+      if (componentStillMounted) {
+        setCollaboratorEmailsStr(collaboratorEmailArr.join(', '));
+      }
+    }
+
+    fetchCollaboratorEmails();
+    return () => { componentStillMounted = false; };
```
Comment (Hiram-Silvey) [inline frontend/src/components/ViewTrips/trip.js:69]: Same as the comment I left on #113.

File: frontend/src/components/Utils/time.test.js (hunk 16 above)
Comment (zghera) [reply frontend/src/components/Utils/time.test.js]: Sure. Fixed.

File: frontend/src/components/ViewTrips/trip.js (hunk 17 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trip.js:70]: Merged up the PR chain from #113. Fixed now.
